### PR TITLE
npm - Partial scan

### PR DESCRIPTION
**Pull request by Or-Geva**
- [x] All [tests](https://github.com/jfrog/jfrog-vscode-extension#building-and-testing-the-sources) passed. If this feature is not already covered by the tests, I added new tests.
- [x] I used `npm run format` for formatting the code before submitting the pull request.
-----
This PR brings an enhancement to the scanning for npm projects. It now enables the scanning process even if there are issues with `npm ls`, such as peer dependencies (as mentioned in this GitHub issue: https://github.com/jfrog/jfrog-vscode-extension/issues/191). The scanning process will do its best to display as much data as possible, from the provided `npm ls` output.

### CODE REVIEW

File: src/main/treeDataProviders/dependenciesTree/dependenciesRoot/npmTree.ts [hunk 1]
```diff
@@ -21,19 +21,27 @@ export class NpmTreeNode extends RootNode {
...
         }
         this.populateDependenciesTree(this, projectDetails.dependencies);
         if (npmLsFailed) {
-            this.topSeverity = Severity.Unknown;
-            this.buildError = BuildTreeErrorType.NotInstalled;
-            this._logManager.logMessageAndToastErr(
-                `Failed to scan npm project. Hint: Please make sure the commands 'npm install' or 'npm ci' run successfully in '${this.workspaceFolder}'`,
-                'ERR'
-            );
+            if (this.children.length === 0){
```
Comment (eyalbe4) [inline src/main/treeDataProviders/dependenciesTree/dependenciesRoot/npmTree.ts:32]: I haven't reviewed this deeply, but just noticed the missing white space before the {
Further review is still required.

File: src/main/treeDataProviders/dependenciesTree/dependenciesRoot/npmTree.ts [hunk 2]
```diff
@@ -23,17 +23,25 @@ export class NpmTreeNode extends RootNode {
...
-            );
+            if (this.children.length === 0) {
+                this.topSeverity = Severity.Unknown;
+                this.buildError = BuildTreeErrorType.NotInstalled;
+                this._logManager.logMessageAndToastErr(
+                    `Failed to scan npm project. Hint: Please make sure the commands 'npm install' or 'npm ci' run successfully in '${this.workspaceFolder}'`,
+                    'ERR'
+                );
+            } else {
+                this._logManager.logMessageAndToastErr(
```
Comment (yahavi) [inline src/main/treeDataProviders/dependenciesTree/dependenciesRoot/npmTree.ts:40]: I think toasting is too aggressive here. There are projects that have errors but still wish to utilize this plugin without addressing their errors.
```suggestion
                this._logManager.logMessage(
```

File: src/main/treeDataProviders/dependenciesTree/dependenciesRoot/npmTree.ts [hunk 3]
```diff
@@ -23,17 +23,25 @@ export class NpmTreeNode extends RootNode {
             projectDetails.loadProjectDetails(NpmCmd.runNpmLs(this.workspaceFolder));
         } catch (error) {
             this._logManager.logError(<any>error, false);
+            projectDetails.loadProjectDetails(JSON.parse((<any>error).stdout.toString()));
```
Comment (yahavi) [inline src/main/treeDataProviders/dependenciesTree/dependenciesRoot/npmTree.ts:26]: Looks like `loadProjectDetailsFromFile` overrides the package name and version assignment you did in loadProjectDetails.

Please consider the following:
```typescript
        let npmLsResult: string;
        try {
            npmLsResult = NpmCmd.runNpmLs(this.workspaceFolder);
        } catch (error) {
            this._logManager.logError(<any>error, false);
            npmLsResult = JSON.parse((<any>error).stdout.toString());
            npmLsFailed = true;
        }
        projectDetails.loadProjectDetails(npmLsResult);
```

File: src/main/treeDataProviders/dependenciesTree/dependenciesRoot/npmTree.ts [hunk 4]
```diff
@@ -23,17 +23,25 @@ export class NpmTreeNode extends RootNode {
...
+            if (this.children.length === 0) {
+                this.topSeverity = Severity.Unknown;
+                this.buildError = BuildTreeErrorType.NotInstalled;
+                this._logManager.logMessageAndToastErr(
+                    `Failed to scan npm project. Hint: Please make sure the commands 'npm install' or 'npm ci' run successfully in '${this.workspaceFolder}'`,
+                    'ERR'
+                );
+            } else {
+                this._logManager.logMessageAndToastErr(
+                    `The npm project was partially scanned. Please ensure that there are no errors from the command 'npm ls' in the directory '${this.workspaceFolder}''`,
```
Comment (yahavi) [inline src/main/treeDataProviders/dependenciesTree/dependenciesRoot/npmTree.ts:41]:
```suggestion
                    `An npm project was partially scanned. Hint: Please ensure that there are no errors from the command 'npm ls --all' in the directory '${this.workspaceFolder}''`,
```

**Review by yahavi — CHANGES_REQUESTED** [review]
Great initiative @Or-Geva! I liked the greedy approach.
Please consider my comments.

**Pull request merged**